// @skyway-sdk/core/dist/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextConfig = void 0;
const deepmerge_1 = __importDefault(require("deepmerge"));
class ContextConfig {
    /**@internal */
    constructor(options = {}) {
        /**@internal */
        this.rtcApi = {
            domain: 'rtc-api.skyway.ntt.com',
            timeout: 30000,
            secure: true,
            eventSubscribeTimeout: 5000,
        };
        /**@internal */
        this.iceParamServer = {
            domain: 'ice-params.skyway.ntt.com',
            version: 1,
            secure: true,
        };
        /**@internal */
        this.signalingService = {
            domain: 'signaling.skyway.ntt.com',
            secure: true,
        };
        /**@internal */
        this.analyticsService = {
            domain: 'analytics-logging.skyway.ntt.com',
            secure: true,
        };
        this.rtcConfig = {
            timeout: 30000,
            turnPolicy: 'enable',
            turnProtocol: 'all',
            encodedInsertableStreams: false,
            iceDisconnectBufferTimeout: 5000,
        };
        this.token = {
            updateReminderSec: 30,
        };
        this.log = {
            level: 'error',
            format: 'string',
        };
        /**@internal */
        this.internal = {
            disableDPlane: false,
        };
        this.member = {
            keepaliveIntervalGapSec: 30,
            keepaliveIntervalSec: 30,
        };
        Object.assign(this, (0, deepmerge_1.default)(this, options));
    }
}
exports.ContextConfig = ContextConfig;
