// @skyway-sdk/room/dist/room/sfu.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SfuRoomImpl = void 0;
const common_1 = require("@skyway-sdk/common");
const sfu_bot_1 = require("@skyway-sdk/sfu-bot");
const sfu_1 = require("../member/local/sfu");
const base_1 = require("../member/remote/base");
const base_2 = require("./base");
const log = new common_1.Logger('packages/room/src/room/sfu.ts');
/**@internal */
class SfuRoomImpl extends base_2.RoomImpl {
    constructor(channel, _plugin) {
        super('sfu', channel);
        this._plugin = _plugin;
        this.setChannelState();
        this.setChannelListener();
    }
    static Create(context, channel) {
        return __awaiter(this, void 0, void 0, function* () {
            const plugin = context.plugins.find((p) => p.subtype === 'sfu');
            const bot = channel.members.find((m) => m.subtype === sfu_bot_1.SfuBotMember.subtype);
            if (!bot) {
                yield plugin.createBot(channel);
            }
            const room = new SfuRoomImpl(channel, plugin);
            return room;
        });
    }
    setChannelState() {
        this._channel.members.forEach((m) => {
            if (m.type === 'bot') {
                return;
            }
            const member = new base_1.RemoteRoomMemberImpl(m, this);
            this._members[m.id] = member;
        });
        this._channel.publications.forEach((p) => {
            if (!p.origin) {
                return;
            }
            this._addPublication(p);
        });
        this._channel.subscriptions.forEach((s) => {
            if (s.subscriber.type === 'bot') {
                return;
            }
            this._addSubscription(s);
        });
    }
    setChannelListener() {
        this._channel.onMemberJoined.add((e) => this._handleOnMemberJoin(e.member));
        this._channel.onMemberLeft.add((e) => this._handleOnMemberLeft(e.member));
        this._channel.onStreamPublished.add((e) => {
            this._handleOnStreamPublish(e.publication);
        });
        this._channel.onStreamUnpublished.add((e) => this._handleOnStreamUnpublish(e.publication));
        this._channel.onPublicationMetadataUpdated.add((e) => {
            this._handleOnPublicationMetadataUpdate(e.publication);
        });
        this._channel.onPublicationEnabled.add((e) => {
            this._handleOnPublicationEnabled(e.publication);
        });
        this._channel.onPublicationDisabled.add((e) => {
            this._handleOnPublicationDisabled(e.publication);
        });
        this._channel.onPublicationSubscribed.add((e) => {
            this._handleOnStreamSubscribe(e.subscription);
        });
        this._channel.onPublicationUnsubscribed.add((e) => this._handleOnStreamUnsubscribe(e.subscription));
    }
    _handleOnMemberJoin(m) {
        if (m.type === 'bot') {
            return;
        }
        const member = new base_1.RemoteRoomMemberImpl(m, this);
        this._members[m.id] = member;
        this.onMemberJoined.emit({ member });
        this.onMemberListChanged.emit({});
    }
    _handleOnMemberLeft(m) {
        const member = this._getMember(m.id);
        if (!member) {
            // should be sfu
            return;
        }
        delete this._members[m.id];
        member._dispose();
        this.onMemberLeft.emit({ member });
        this.onMemberListChanged.emit({});
    }
    _handleOnStreamPublish(p) {
        var _a;
        if (!((_a = p.origin) === null || _a === void 0 ? void 0 : _a.id)) {
            return;
        }
        const publication = this._addPublication(p);
        this.onStreamPublished.emit({ publication });
        this.onPublicationListChanged.emit({});
    }
    _handleOnStreamUnpublish(p) {
        var _a;
        if (!((_a = p.origin) === null || _a === void 0 ? void 0 : _a.id)) {
            return;
        }
        const publication = this._getPublication(p.id);
        delete this._publications[p.id];
        this.onStreamUnpublished.emit({ publication });
        this.onPublicationListChanged.emit({});
    }
    _getRelayedPublication(publicationId) {
        const relayed = this.publications.find((p) => { var _a; return ((_a = p._publication.origin) === null || _a === void 0 ? void 0 : _a.id) === publicationId; });
        return relayed;
    }
    _handleOnPublicationMetadataUpdate(p) {
        const publication = this._getRelayedPublication(p.id);
        if (!publication)
            return;
        this.onPublicationMetadataUpdated.emit({
            publication,
            metadata: publication.metadata,
        });
    }
    _handleOnPublicationEnabled(p) {
        const publication = this._getRelayedPublication(p.id);
        if (!publication)
            return;
        this.onPublicationEnabled.emit({ publication });
    }
    _handleOnPublicationDisabled(p) {
        const publication = this._getRelayedPublication(p.id);
        if (!publication)
            return;
        this.onPublicationDisabled.emit({ publication });
    }
    _handleOnStreamSubscribe(s) {
        if (s.subscriber.type === 'bot') {
            return;
        }
        const subscription = this._addSubscription(s);
        this.onPublicationSubscribed.emit({ subscription });
        this.onSubscriptionListChanged.emit({});
    }
    _handleOnStreamUnsubscribe(s) {
        if (s.subscriber.type === 'bot') {
            return;
        }
        const subscription = this._getSubscription(s.id);
        delete this._subscriptions[s.id];
        this.onPublicationUnsubscribed.emit({ subscription });
        this.onSubscriptionListChanged.emit({});
    }
    join(memberInit = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const local = yield this.joinChannel(Object.assign(Object.assign({}, memberInit), { disableSignaling: true }));
            const localRoomMember = new sfu_1.LocalSFURoomMemberImpl(local, this);
            this.localRoomMember = localRoomMember;
            localRoomMember.onLeft.once(() => {
                this.localRoomMember = undefined;
            });
            log.debug('member joined', memberInit);
            return localRoomMember;
        });
    }
}
exports.SfuRoomImpl = SfuRoomImpl;
