// @skyway-sdk/room/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamFactory = exports.SkyWayStreamFactory = exports.SkyWayContext = exports.RemoteVideoStream = exports.RemoteStreamBase = exports.RemoteMediaStreamBase = exports.RemoteDataStream = exports.RemoteAudioStream = exports.MediaDevice = exports.LocalVideoStream = exports.LocalStreamBase = exports.LocalMediaStreamBase = exports.LocalDataStream = exports.LocalCustomVideoStream = exports.LocalAudioStream = exports.getRtcRtpCapabilities = exports.getBitrateFromPeerConnection = exports.Events = exports.Event = exports.createTestVideoTrack = exports.ContextConfig = exports.errors = void 0;
var errors_1 = require("./errors");
Object.defineProperty(exports, "errors", { enumerable: true, get: function () { return errors_1.errors; } });
__exportStar(require("./member"), exports);
__exportStar(require("./member/local/base"), exports);
__exportStar(require("./member/local/p2p"), exports);
__exportStar(require("./member/local/sfu"), exports);
__exportStar(require("./member/remote/base"), exports);
__exportStar(require("./publication"), exports);
__exportStar(require("./room"), exports);
__exportStar(require("./room/base"), exports);
__exportStar(require("./room/event"), exports);
__exportStar(require("./room/p2p"), exports);
__exportStar(require("./room/sfu"), exports);
__exportStar(require("./subscription"), exports);
__exportStar(require("./version"), exports);
__exportStar(require("@skyway-sdk/common"), exports);
var core_1 = require("@skyway-sdk/core");
Object.defineProperty(exports, "ContextConfig", { enumerable: true, get: function () { return core_1.ContextConfig; } });
Object.defineProperty(exports, "createTestVideoTrack", { enumerable: true, get: function () { return core_1.createTestVideoTrack; } });
Object.defineProperty(exports, "Event", { enumerable: true, get: function () { return core_1.Event; } });
Object.defineProperty(exports, "Events", { enumerable: true, get: function () { return core_1.Events; } });
Object.defineProperty(exports, "getBitrateFromPeerConnection", { enumerable: true, get: function () { return core_1.getBitrateFromPeerConnection; } });
Object.defineProperty(exports, "getRtcRtpCapabilities", { enumerable: true, get: function () { return core_1.getRtcRtpCapabilities; } });
Object.defineProperty(exports, "LocalAudioStream", { enumerable: true, get: function () { return core_1.LocalAudioStream; } });
Object.defineProperty(exports, "LocalCustomVideoStream", { enumerable: true, get: function () { return core_1.LocalCustomVideoStream; } });
Object.defineProperty(exports, "LocalDataStream", { enumerable: true, get: function () { return core_1.LocalDataStream; } });
Object.defineProperty(exports, "LocalMediaStreamBase", { enumerable: true, get: function () { return core_1.LocalMediaStreamBase; } });
Object.defineProperty(exports, "LocalStreamBase", { enumerable: true, get: function () { return core_1.LocalStreamBase; } });
Object.defineProperty(exports, "LocalVideoStream", { enumerable: true, get: function () { return core_1.LocalVideoStream; } });
Object.defineProperty(exports, "MediaDevice", { enumerable: true, get: function () { return core_1.MediaDevice; } });
Object.defineProperty(exports, "RemoteAudioStream", { enumerable: true, get: function () { return core_1.RemoteAudioStream; } });
Object.defineProperty(exports, "RemoteDataStream", { enumerable: true, get: function () { return core_1.RemoteDataStream; } });
Object.defineProperty(exports, "RemoteMediaStreamBase", { enumerable: true, get: function () { return core_1.RemoteMediaStreamBase; } });
Object.defineProperty(exports, "RemoteStreamBase", { enumerable: true, get: function () { return core_1.RemoteStreamBase; } });
Object.defineProperty(exports, "RemoteVideoStream", { enumerable: true, get: function () { return core_1.RemoteVideoStream; } });
Object.defineProperty(exports, "SkyWayContext", { enumerable: true, get: function () { return core_1.SkyWayContext; } });
Object.defineProperty(exports, "SkyWayStreamFactory", { enumerable: true, get: function () { return core_1.SkyWayStreamFactory; } });
Object.defineProperty(exports, "StreamFactory", { enumerable: true, get: function () { return core_1.StreamFactory; } });
__exportStar(require("@skyway-sdk/token"), exports);
