// @skyway-sdk/room/dist/member/local/sfu.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalSFURoomMemberImpl = void 0;
const common_1 = require("@skyway-sdk/common");
const core_1 = require("@skyway-sdk/core");
const sfu_bot_1 = require("@skyway-sdk/sfu-bot");
const const_1 = require("../../const");
const errors_1 = require("../../errors");
const util_1 = require("../../util");
const base_1 = require("./base");
const log = new common_1.Logger('packages/room/src/member/local/sfu.ts');
/**@internal */
class LocalSFURoomMemberImpl extends base_1.LocalRoomMemberImpl {
    /**@private */
    constructor(member, room) {
        super(member, room);
    }
    publish(stream, options = {}) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (stream instanceof core_1.LocalDataStream) {
                throw (0, util_1.createError)({
                    operationName: 'LocalSFURoomMemberImpl.publish',
                    context: this._context,
                    room: this.room,
                    info: errors_1.errors.sfuRoomNotSupportDataStream,
                    path: log.prefix,
                });
            }
            options.maxSubscribers = (_a = options.maxSubscribers) !== null && _a !== void 0 ? _a : const_1.defaultMaxSubscribers;
            const origin = yield this._local.publish(stream, options);
            const bot = this.room._channel.members.find((m) => m.subtype === sfu_bot_1.SfuBotMember.subtype);
            if (!bot) {
                throw (0, util_1.createError)({
                    operationName: 'LocalSFURoomMemberImpl.publish',
                    context: this._context,
                    room: this.room,
                    info: sfu_bot_1.errors.sfuBotNotInChannel,
                    path: log.prefix,
                });
            }
            const forwarding = yield bot.startForwarding(origin, {
                maxSubscribers: options.maxSubscribers,
            });
            const relayingPublication = forwarding.relayingPublication;
            const roomPublication = this.room._addPublication(relayingPublication);
            this.onStreamPublished.emit({ publication: roomPublication });
            return roomPublication;
        });
    }
    /**
     * @description [japanese] Room上のStreamをUnPublishする
     */
    unpublish(target) {
        return __awaiter(this, void 0, void 0, function* () {
            const publicationId = typeof target === 'string' ? target : target.id;
            const publication = this.room._getPublication(publicationId);
            const origin = publication._publication.origin;
            if (!origin) {
                throw (0, util_1.createError)({
                    operationName: 'LocalSFURoomMemberImpl.unpublish',
                    context: this._context,
                    room: this.room,
                    info: errors_1.errors.publicationNotHasOrigin,
                    path: log.prefix,
                });
            }
            this._local.unpublish(origin.id).catch((error) => {
                log.error('unpublish error', error, { target }, this.toJSON());
            });
            yield this.room.onStreamUnpublished
                .watch((e) => e.publication.id === publicationId, this._context.config.rtcApi.timeout)
                .catch((error) => {
                throw (0, util_1.createError)({
                    operationName: 'LocalSFURoomMemberImpl.unpublish',
                    context: this._context,
                    room: this.room,
                    info: Object.assign(Object.assign({}, errors_1.errors.timeout), { detail: 'onStreamUnpublished' }),
                    path: log.prefix,
                    error,
                });
            });
            this.onStreamUnpublished.emit({ publication });
        });
    }
    /**
     * @description [japanese] MemberがRoom上のStreamのPublicationをSubscribeする
     */
    subscribe(target, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const publicationId = typeof target === 'string' ? target : target.id;
            const { subscription, stream } = yield this._local.subscribe(publicationId, options);
            const roomSubscription = this.room._addSubscription(subscription);
            return {
                subscription: roomSubscription,
                stream: stream,
            };
        });
    }
    /**
     * @description [japanese] MemberがSubscribeしているStreamのSubscriptionをUnSubscribeする
     */
    unsubscribe(target) {
        return __awaiter(this, void 0, void 0, function* () {
            const subscriptionId = typeof target === 'string' ? target : target.id;
            this._local.unsubscribe(subscriptionId).catch((error) => {
                log.error('unsubscribe error', error, { target }, this.toJSON());
            });
            yield this.room.onPublicationUnsubscribed
                .watch((e) => e.subscription.id === subscriptionId, this._context.config.rtcApi.timeout)
                .catch((error) => {
                throw (0, util_1.createError)({
                    operationName: 'LocalSFURoomMemberImpl.unsubscribe',
                    context: this._context,
                    room: this.room,
                    info: Object.assign(Object.assign({}, errors_1.errors.timeout), { detail: 'onPublicationUnsubscribed' }),
                    path: log.prefix,
                    error,
                });
            });
        });
    }
    _updateRoom(room) {
        log.debug('_updateRoom', { memberId: this.id });
        this.room = room;
        this._listenRoomEvent();
    }
}
exports.LocalSFURoomMemberImpl = LocalSFURoomMemberImpl;
