// @skyway-sdk/room/dist/member/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoomMemberImpl = void 0;
const common_1 = require("@skyway-sdk/common");
/**@internal */
class RoomMemberImpl {
    constructor(member, room) {
        this.member = member;
        this.room = room;
        this.onLeft = new common_1.Event();
        const { removeListener } = room.onMemberLeft.add((e) => {
            if (e.member.id === this.member.id) {
                removeListener();
                this.onLeft.emit();
            }
        });
        this.onMetadataUpdated = member.onMetadataUpdated;
    }
    get id() {
        return this.member.id;
    }
    get name() {
        return this.member.name;
    }
    get roomId() {
        return this.room.id;
    }
    get roomName() {
        return this.room.name;
    }
    get roomType() {
        return this.room.type;
    }
    get state() {
        return this.member.state;
    }
    get metadata() {
        return this.member.metadata;
    }
    /**@private */
    get _member() {
        return this.member;
    }
    get publications() {
        return this.room.publications.filter((p) => p.publisher.id === this.id);
    }
    get subscriptions() {
        return this.member.subscriptions.map((s) => this.room._getSubscription(s.id));
    }
    updateMetadata(metadata) {
        return this.member.updateMetadata(metadata);
    }
    leave() {
        return this.member.leave();
    }
    /**@internal */
    toJSON() {
        return { id: this.id, name: this.name, metadata: this.metadata };
    }
}
exports.RoomMemberImpl = RoomMemberImpl;
