// @skyway-sdk/core/dist/media/stream/local/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalVideoStream = exports.LocalStreamBase = exports.LocalMediaStreamBase = exports.LocalDataStream = exports.LocalCustomVideoStream = exports.LocalAudioStream = void 0;
const audio_1 = require("./audio");
Object.defineProperty(exports, "LocalAudioStream", { enumerable: true, get: function () { return audio_1.LocalAudioStream; } });
const base_1 = require("./base");
Object.defineProperty(exports, "LocalStreamBase", { enumerable: true, get: function () { return base_1.LocalStreamBase; } });
const customVideo_1 = require("./customVideo");
Object.defineProperty(exports, "LocalCustomVideoStream", { enumerable: true, get: function () { return customVideo_1.LocalCustomVideoStream; } });
const data_1 = require("./data");
Object.defineProperty(exports, "LocalDataStream", { enumerable: true, get: function () { return data_1.LocalDataStream; } });
const media_1 = require("./media");
Object.defineProperty(exports, "LocalMediaStreamBase", { enumerable: true, get: function () { return media_1.LocalMediaStreamBase; } });
const video_1 = require("./video");
Object.defineProperty(exports, "LocalVideoStream", { enumerable: true, get: function () { return video_1.LocalVideoStream; } });
