// @skyway-sdk/sfu-bot/dist/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createWarnPayload = exports.moveToHead = exports.getLayerFromEncodings = void 0;
function getLayerFromEncodings(id, encodings) {
    let layer = 0;
    for (; layer < encodings.length; layer++) {
        const encoding = encodings[layer];
        if (encoding.id === id) {
            break;
        }
    }
    return layer;
}
exports.getLayerFromEncodings = getLayerFromEncodings;
function moveToHead(arr, selector) {
    const target = arr.find(selector);
    return [target, ...arr.filter((o) => !selector(o))];
}
exports.moveToHead = moveToHead;
function createWarnPayload({ channel, detail, operationName, payload, bot, }) {
    var _a;
    const warn = {
        operationName,
        payload,
        detail,
    };
    if (channel) {
        warn['appId'] = channel.appId;
        warn['channelId'] = channel.id;
        if (channel.localPerson) {
            warn['memberId'] = channel.localPerson.id;
        }
    }
    if (bot) {
        warn['botId'] = bot.id;
        warn['appId'] = bot.channel.appId;
        warn['channelId'] = bot.channel.id;
        warn['memberId'] = (_a = bot.channel.localPerson) === null || _a === void 0 ? void 0 : _a.id;
    }
    return warn;
}
exports.createWarnPayload = createWarnPayload;
